naclbox/keysource.go

```diff
@@ -0,0 +1,219 @@
+package naclbox //import "go.mozilla.org/sops/naclbox"
+
+import (
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"go.mozilla.org/sops/logging"
+	"golang.org/x/crypto/nacl/box"
+)
+
+var log *logrus.Logger
+
+func init() {
+	log = logging.NewLogger("NACLBOX")
+}
+
+// MasterKey is a NACL key used to encrypt and decrypt sops' data key.
+type MasterKey struct {
+	PublicKey       string    // a base64 encoded public key
+	pub, priv       *[32]byte // decoded public and private nacl box keys
+	EncryptedKey    string
+	EphemeralPubKey string // base64 encoded public key used for authentication
+	Nonce           string // base64 encoded unique nonce
+	CreationDate    time.Time
+}
+
+// KeyFile is the file representation of a public and private NACL BOX keypair
+type KeyFile struct {
+	PublicKey, PrivateKey string
+}
+
+// EncryptedDataKey returns the encrypted data key this master key holds
+func (key *MasterKey) EncryptedDataKey() []byte {
+	return []byte(key.EncryptedKey)
+}
+
+// SetEncryptedDataKey sets the encrypted data key for this master key
+func (key *MasterKey) SetEncryptedDataKey(enc []byte) {
+	key.EncryptedKey = string(enc)
+}
+
+// Encrypt takes a sops data key, encrypts it with NACL BOX
+// and stores the result in the EncryptedKey field
+func (key *MasterKey) Encrypt(dataKey []byte) error {
+	// NACL BOX uses two keys pairs to seal secret data: sender and recipient.
+	// When used in Sops, we don't care much about the sender keypair, but it is
+	// required for the protocol to work, so we issue an ephemeral keypair instead
+	// and store its pubkey alongside the encrypted data key, so it can later
+	// be decrypted.
```

Comment:
Should this say "so it can later be reencrypted"? Unless I'm misunderstanding, you're generating a key pair and throwing away the priv key (since it's not going to decrypt anything), and we need to keep it so we can reencrypt the data key in the future if needed without generating a completely new keypair.